File: SourceCode/Python/Binary/RedBlackTree.py
class Node:
    """A node in a red-black tree."""

    def __init__(self, is_red, key, value, left = None, right = None):
        self._is_red = is_red
        self.key = key
        self.value = value
        self.left = left
        self.right = right

    def is_red(self):
        if self is None:
            return False
        else:
            return self._is_red

    def is_black(self):
        return not Node.is_red(self)

    def __str__(self):
        return "Node(%s, %s, %s, %s, %s)" % \
            (self._is_red, self.key, self.value, self.left, self.right)

class RedBlackTree:
    """A dictionary implemented using a binary search tree."""

    def __init__(self):
        self.root = None

    def check_invariant(self):
        """Check that the invariant holds."""

        if Node.is_red(self.root):
            raise AssertionError("red root")

        self.check_invariant_recursive(self.root, None, None)

    @staticmethod
    def check_invariant_recursive(node, lo, hi):
        """Helper method for 'check_invariant'.

        Checks that the node is the root of a valid red-black tree, and that
        all keys k satisfy lo < k < hi. The test lo < k is skipped
        if lo is None, and k < hi is skipped if hi is None.

        Returns the "black height" of the tree."""

        if node is None: return 0

        if Node.is_red(node.right):
            raise AssertionError("red right child")

        if Node.is_red(node) and Node.is_red(node.left):
            raise AssertionError("red node with red left child")

        if lo is not None and node.key <= lo:
            raise AssertionError("key too small", node.key, lo, hi)

        if hi is not None and node.key >= hi:
            raise AssertionError("key too big", node.key, lo, hi)

        # Keys in the left subtree should be < node.key
        h1 = RedBlackTree.check_invariant_recursive(node.left, lo, node.key)
        # Keys in the right subtree should be > node.key
        h2 = RedBlackTree.check_invariant_recursive(node.right, node.key, hi)

        if h1 != h2:
            raise AssertionError("unbalanced tree")

        return h1 + (1 if Node.is_black(node) else 0)

    def add(self, key, value):
        """Add a key-value pair, or update the value associated with
        an existing key."""

        self.root = self.add_recursive(self.root, key, value)
        if Node.is_red(self.root):
            self.root._is_red = False

    @staticmethod
    def add_recursive(node, key, value):
        """Helper method for 'add'."""

        if node is None:
            return Node(True, key, value, None, None)
        elif key < node.key:
            node.left = RedBlackTree.add_recursive(node.left, key, value)
        elif key > node.key:
            node.right = RedBlackTree.add_recursive(node.right, key, value)
        else:
            node.value = value
        return RedBlackTree.rebalance(node)

    def get(self, key):
        """Look up a key."""

        return self.get_recursive(self.root, key)

    @staticmethod
    def get_recursive(node, key):
        """Helper method for 'get'."""

        if node is None:
            return None
        elif key < node.key:
            return RedBlackTree.get_recursive(node.left, key)
        elif key > node.key:
            return RedBlackTree.get_recursive(node.right, key)
        else:
            return node.value

    @staticmethod
    def rebalance(node):
        if node is None: return None
        
        # Skew
        if Node.is_red(node.right):
            node = RedBlackTree.rotate_left(node)

        # Split part 1
        if Node.is_red(node.left) and Node.is_red(node.left.left):
            node = RedBlackTree.rotate_right(node)

        # Split part 2
        if Node.is_red(node.left) and Node.is_red(node.right):
            node.left._is_red = False
            node.right._is_red = False
            node._is_red = True

        return node

    @staticmethod
    def rotate_left(node):
        node1 = node.left
        node2 = node.right.left
        node3 = node.right.right
        key1 = node.key
        value1 = node.value
        red1 = node._is_red
        key2 = node.right.key
        value2 = node.right.value
        red2 = node.right._is_red

        return Node(red1, key2, value2, Node(red2, key1, value1, node1, node2), node3)

    @staticmethod
    def rotate_right(node):
        node1 = node.left.left
        node2 = node.left.right
        node3 = node.right
        key1 = node.left.key
        value1 = node.left.value
        red1 = node.left._is_red
        key2 = node.key
        value2 = node.value
        red2 = node._is_red

        return Node(red2, key1, value1, node1, Node(red1, key2, value2, node2, node3))

    def __iter__(self):
        """Iterate through all keys.

        This is called when the user writes 'for key in bst: ...'."""

        return self.iter_recursive(self.root)

    @staticmethod
    def iter_recursive(node):
        """Helper method for '__iter__'."""

        # This method is a generator:
        # https://docs.python.org/3/howto/functional.html#generators
        # Generators are an easy way to make iterators

        if node is None:
            return
        else:
            for key in RedBlackTree.iter_recursive(node.left):
                yield key
            yield node.key
            for key in RedBlackTree.iter_recursive(node.right):
                yield key

    def __getitem__(self, key):
        """This is called when the user writes 'x = bst[key]'."""

        return self.get(key)
    
    def __setitem__(self, key, value):
        """This is called when the user writes 'bst[key] = value'."""

        self.add(key, value)

    def __delitem__(self, key):
        """This is called when the user writes 'del bst[key]'."""

        self.delete(key)

    def __str__(self):
        """This is called to show the red-black tree as a string."""

        # Use a dict comprehension to convert the red-black tree into a dict:
        # https://docs.python.org/3/tutorial/datastructures.html#dictionaries
        # Then show the dict as a string.

        return str({key: self[key] for key in self})
        # This code is the same as:
        # d = {}
        # for key in self:
        #   d[key] = self[key]
        # return str(d)
        # Note that 'for key in self' calls self.__iter__ to produce
        # the keys, and 'self[key]' calls self.__getitem__.

    def __repr__(self):
        """This is called to show the red-black tree as a string."""

        return repr({key: self[key] for key in self})

if __name__ == '__main__':
    bst = RedBlackTree()
    keys = [3,1,4,1,5,9,2,6,5,3,5,8,9,7,9,3,2,3,8,4,6]
    values = list(range(len(keys)))

    for i in range(len(keys)):
        bst[keys[i]] = values[i]
        print(bst.root)
        bst.check_invariant()
    for i in range(len(keys)):
        print(bst[keys[i]])
        bst.check_invariant()

File: SourceCode/Java_Generic/Binary/RedBlackTree.java
import java.util.ArrayList;
import java.util.Iterator;

/* *** ODSATag: RedBlackTree *** */
// A dictionary implemented using an red-black tree.
public class RedBlackTree<Key extends Comparable<Key>, Value> implements Iterable<Key> {
    // A node in an red-black tree.
    class Node {
        Key key;
        Value value;
        Node left;
        Node right;
        boolean isRed;

        Node(boolean isRed, Key key, Value value, Node left, Node right) {
            this.key = key;
            this.value = value;
            this.left = left;
            this.right = right;
            this.isRed = isRed;
        }
    }

    // Check if a node is red. 'null' is always black.
    boolean isRed(Node node) {
        if (node == null) return false;
        return node.isRed;
    }

    // The root of the binary search tree.
    Node root = null;

    // Check that the invariant holds.
    void checkInvariant() {
        if (isRed(root))
            throw new AssertionError("red root");

        checkInvariantHelper(root, null, null);
    }

    // Helper method for 'check_invariant'.
    //
    // Checks that the node is the root of a valid red-black tree, and that
    // all keys k satisfy lo < k < hi. The test lo < k is skipped
    // if lo is None, and k < hi is skipped if hi is None.
    //
    // Returns the "black height" of the tree.
    int checkInvariantHelper(Node node, Key lo, Key hi) {
        if (node == null) return 0;

        if (isRed(node.right))
            throw new AssertionError("red right child");

        if (isRed(node) && isRed(node.left))
            throw new AssertionError("red node with red left child");

        if (lo != null && node.key.compareTo(lo) <= 0)
            throw new AssertionError("key too small");

        if (hi != null && node.key.compareTo(hi) >= 0)
            throw new AssertionError("key too big");

        // Keys in the left subtree should be < node.key
        int h1 = checkInvariantHelper(node.left, lo, node.key);
        // Keys in the right subtree should be > node.key
        int h2 = checkInvariantHelper(node.right, node.key, hi);

        if (h1 != h2)
            throw new AssertionError("unbalanced tree");

        return h1 + (isRed(node) ? 0 : 1);
    }

    // Add a key-value pair, or update the value associated with an existing key.
    public void add(Key key, Value value) {
        root = addHelper(root, key, value);
        if (isRed(root)) root.isRed = false;
    }

    // Helper method for 'add'.
    Node addHelper(Node node, Key key, Value value) {
        if (node == null)
            return new Node(true, key, value, null, null);

        else if (key.compareTo(node.key) < 0)
            node.left = addHelper(node.left, key, value);

        else if (key.compareTo(node.key) > 0)
            node.right = addHelper(node.right, key, value);

        else
            node.value = value;

        return rebalance(node);
    }

    // Look up a key.
    public Value get(Key key) {
        return getHelper(root, key);
    }

    // Helper method for 'get'.
    Value getHelper(Node node, Key key) {
        if (node == null)
            return null;

        else if (key.compareTo(node.key) < 0)
            return getHelper(node.left, key);

        else if (key.compareTo(node.key) > 0)
            return getHelper(node.right, key);

        else
            return node.value;
    }

    // Repair the red-black invariant by rebalancing the node.
    Node rebalance(Node node) {
        if (node == null) return node;

        // Skew
        if (isRed(node.right))
            node = rotateLeft(node);

        // Split part 1
        if (isRed(node.left) && isRed(node.left.left))
            node = rotateRight(node);

        // Split part 2
        if (isRed(node.left) && isRed(node.right)) {
            node.left.isRed = false;
            node.right.isRed = false;
            node.isRed = true;
        }

        return node;
    }

    Node rotateLeft(Node node) {
        // Left rotation.

        //    x                y
        //   / \              / \
        //  A  y      ===>   x  C
        //    / \           / \
        //   B  C          A  B
        // """

        // Variables are named according to the picture above.
        Node x = node;
        Node A = x.left;
        Node y = x.right;
        Node B = y.left;
        Node C = y.right;

        // We also swap x's and y's colours
        // (e.g. if x was black before, then y will be black afterwards).
        return new Node(x.isRed, y.key, y.value, new Node(y.isRed, x.key, x.value, A, B), C);
    }

    Node rotateRight(Node node) {
        // Right rotation.
        // 
        //      x             y
        //     / \           / \
        //    y  C    ===>  A  x
        //   / \              / \
        //  A  B             B  C

        // Variables are named according to the picture above.
        Node x = node;
        Node y = x.left;
        Node A = y.left;
        Node B = y.right;
        Node C = x.right;

        // We also swap x's and y's colours
        // (e.g. if x was black before, then y will be black afterwards).
        return new Node(x.isRed, y.key, y.value, A, new Node(y.isRed, x.key, x.value, B, C));
    }

    // Iterate through all keys.
    // This is called when the user writes 'for (Key key: bst) { ... }.'
    public Iterator<Key> iterator() {
        // The easiest way to solve this is to add all keys to an
        // ArrayList, then iterate through that.
        ArrayList<Key> keys = new ArrayList<>();
        iteratorHelper(root, keys);
        return keys.iterator();
    }

    // Helper method for 'iterator'
    void iteratorHelper(Node node, ArrayList<Key> keys) {
        if (node == null) return;
        iteratorHelper(node.left, keys);
        keys.add(node.key);
        iteratorHelper(node.right, keys);
    }

    // Override 'toString' to print the contents of the red-black tree.
    public String toString() {
        StringBuilder str = new StringBuilder();
        boolean firstKey = true;

        for (Key key: this) {
            Value value = this.get(key);

            if (!firstKey) str.append(", ");
            str.append(key.toString() + "->" + value.toString());
            firstKey = false;
        }

        return "{" + str + "}";
    }

    // Some test code to check that the red-black tree is working
    public static void main(String[] args) {
        RedBlackTree<Integer, Integer> bst = new RedBlackTree<>();
        int[] keys = {3,1,4,1,5,9,2,6,5,3,5,8,9,7,9,3,2,3,8,4,6};
        int[] values = new int[keys.length];
        for (int i = 0; i < values.length; i++) values[i] = i;

        for (int i = 0; i < keys.length; i++) {
            bst.add(keys[i], values[i]);
            System.out.println(bst);
            bst.checkInvariant();
        }

        for (int i = 0; i < keys.length; i++) {
            System.out.println(bst.get(keys[i]));
            bst.checkInvariant();
        }
    }
}
/* *** ODSAendTag: RedBlackTree *** */
